tests/testmplex.nim

```diff
@@ -140,6 +144,48 @@ suite "Mplex":
       let mplexDial = newMplex(conn)
       let stream  = await mplexDial.newStream()
       await stream.writeLp("Hello from stream!")
+      await lock or timeout
+      check lock.finished
+      timeout.cancel()
+      await conn.close()
+      result = true
+
+    check:
+      waitFor(testNewStream()) == true
+
+  test "e2e - write limits":
+    proc testNewStream(): Future[bool] {.async.} =
+      let
+        ma: MultiAddress = Multiaddress.init("/ip4/0.0.0.0/tcp/0")
+        lock = newFuture[void]()
+        timeout = sleepAsync(5_000)
+
+      proc connHandler(conn: Connection) {.async, gcsafe.} =
+        proc handleMplexListen(stream: Connection) {.async, gcsafe.} =
+          let msg = await stream.readLp()
+          lock.complete()
+          check cast[string](msg) == "Hello from stream!"
+          await stream.close()
+
+        let mplexListen = newMplex(conn)
+        mplexListen.streamHandler = handleMplexListen
+        discard mplexListen.handle()
+
+      let transport1: TcpTransport = newTransport(TcpTransport)
+      discard await transport1.listen(ma, connHandler)
+
+      let transport2: TcpTransport = newTransport(TcpTransport)
+      let conn = await transport2.dial(transport1.ma)
+
+      let mplexDial = newMplex(conn)
+      let stream  = await mplexDial.newStream()
+      var bigseq = newSeqOfCap[uint8](MaxMsgSize + 1)
+      for _ in 0..<MaxMsgSize:
+        bigseq.add(uint8(rand(int('A')..int('z'))))
+      await stream.writeLp(bigseq)
+      await lock or timeout
+      check timeout.finished # this test has to timeout!
+      lock.cancel();
```

Comment:
since the topic of exception safety was raised, this is a school-book example of broken code - on any exception (and there are plenty of them above, even if there's no visible hint of it), this cancel and the close below will not be called - in tests this is annoying because you often get a cascade of failures as ports are not closed correctly, but more generally it's bad because of the proliferation of poor code examples that later are copy-pasted

raising exceptions in nim is very easy and convenient - looking at the outcomes in our code, handling them correctly and consistently turns out to be very hard in practice

Reply:
Well you are right but I see no fault in the way the tests are written. 
This is more because there is no RAII/Destructors/Defer taking care of the situation behind the scene.

Altho isn't this supposed to take care of the scenario you gave? If not I guess we should try/catch/finally every test.
https://github.com/status-im/nim-libp2p/blob/b3f74ba91c23510242be178bf57d08f6665eb5b7/tests/testmplex.nim#L176-L177

Anyway in tests you want to have early failure with constructs like `check`

Reply:
> This is more because there is no RAII/Destructors/Defer taking care of the situation behind the scene.

well, that's the point kind of - if you want to use exceptions in nim correctly, you have to write `try`/`finally`/`defer` - also in tests - else the tests leak resources and are no longer independent - unfortunately, this kind of code is not unusual at all (in fact, both our and standard library code is riddled with leaks like these), tests or "production" code alike. 

In this particular case, the right thing would be be `defer: close` - as for the cancellation, I don't remember if future supports cancelling already-triggered futures - if not, it's defer+if and a flag..

Reply:
Yeah I was thinking to use `defer` but wasn't `defer` broken 😄 ? I refrained using it kinda.. maybe it's fine now.
Anyway I will have a look at it!

Reply:
it used to be - specially in `async`, and still is, but the situation has improved and some simple cases are handled correctly now :)